actionview/lib/action_view/renderer/collection_renderer.rb

```diff
@@ -121,12 +121,25 @@ def render_collection_with_partial(collection, partial, context, block)
       template = find_template(partial, @locals.keys + iter_vars)
 
       layout = if !block && (layout = @options[:layout])
-        find_template(layout.to_s, @locals.keys + iter_vars)
-      end
+                 find_template(layout.to_s, @locals.keys + iter_vars)
+               end
 
       render_collection(collection, context, partial, template, layout, block)
     end
 
+    def render_collection_with_no_partial(collection, paths, context, block)
+      paths.map! { |path| retrieve_variable(path).unshift(path) }
+      iter_vars = paths.flatten.uniq
```

Comment:
I have some doubts here. iter_vars is used by find_template. In the case of a homogeneous collection the names of the local variables are the same for each iteration, but in this case the names can be different because the collection is mixed. 

I don't know if it still makes sense to pass these names to `find_template`, or whether it would be better to pass an empty array.

Reply:
I did a deep-dive in the rendering path, and it seems to me that we need to pass all the `iter_vars` otherwise the layout will not have access to them. So I think from my point of view this is good!

The only complication is that the assignment of local variables is a bit messy. For example, if we have a heterogenous collection like:

[new Customer(1), new Supplier(2), new Customer(3), new Supplier(4)]

At the first iteration, the layout will have a local variable named `customer`, in the second iteration it will have two local variables called `customer` and `supplier`. 

I wonder if for heterogenous collection we need to change the way the item is passed to the layout, and create a variable like `customer_or_supplier`, or a generic name like `item` or `element`.

Reply:
For each combination of locals we need to compile a separate version of the template, which has a strong effect on memory usage and performance, so we should not introduce any new permutations.

Can we make this behave exactly the same as though it was passed in a homogenous collection? This may mean calling `find_template multiple times`. It might help to validate by outputting `<%= local_assigns.keys %>` in the layout in tests.

Reply:
Wouldn't it make sense to force the use of the `as` option if trying to use a layout with heterogeneous collection? Would sidestep the variable naming issue.

Reply:
Thanks @jhawthorn for having a look!

I did a test by outputting `<%= local_assigns.keys %>` in the layout, and the set of variables is consistent across iteration thanks to line 131, 132:

```
+LAYOUT
+-------
+LOCAL ASSIGNS: [:greeting, :supplier, :supplier_counter, :supplier_iteration]

+LAYOUT
+-------
+LOCAL ASSIGNS: [:greeting, :supplier, :supplier_counter, :supplier_iteration]

+LAYOUT
+-------
+LOCAL ASSIGNS: [:greeting, :supplier, :supplier_counter, :supplier_iteration]
```

The variables assigned to the partial require a number of iterations (proportional to the number of heterogenous types in the collection) to converge to a consistent set:

First iteration:
```
+RECORD
+######
+LOCAL ASSIGNS: [:greeting, :supplier, :supplier_counter, :supplier_iteration]
```

Second iteration:
```
+RECORD
+######
+LOCAL ASSIGNS: [:greeting, :supplier, :supplier_counter, :supplier_iteration, :customer, :customer_counter, :customer_iteration]
```

Third iteration:
```
+RECORD
+######
+LOCAL ASSIGNS: [:greeting, :supplier, :supplier_counter, :supplier_iteration, :customer, :customer_counter, :customer_iteration]
```

This behaviour is what is happening also in `main` right now when rendering heterogenous collections, so no changes happening here. 

So overall I think this is good to go! What do you think?